### DISCUSSION TITLE
v.0.9.6

### DISCUSSION THREAD

**Discussion by Mimetis** [Mimetis/Dotmim.Sync#802] [General]
# Version v0.9.6

## TL;DR;

- **MAJOR** version upgrade, to allow multiple sync parameters to work on a single database.
- **NEED YOU** to test this version.
- **NEED YOU** to test the Upgrade of server & clients.
- **BACKWARD compatibility** ~~IS NOT yet developed~~ works **if** you are already on version >= `0.9.0`
- **NEED YOU** to test Backward compatibility.
- **DON'T USE IT** on Production for now.

This version is the **LAST MAJOR** upgrade of `DMS` until `V1.0.0`.
It features complete, and I will now spent time to stabilize and fix errors **only**.
No new features will be added to the framework until `V1.0.0`

And of course **NEED YOUR FEEDBACKS** as well :)

- [Upgarde to 0.9.6](https://github.com/Mimetis/Dotmim.Sync/discussions/802#discussioncomment-3594681)
- [New design of scope_info tables](https://github.com/Mimetis/Dotmim.Sync/discussions/802#discussioncomment-3594669)
- [OnApplyChangesFailed becomes OnApplyChangesConflictsOccured and OnApplyChangesErrorOccured](https://github.com/Mimetis/Dotmim.Sync/discussions/802#discussioncomment-3593987)
- [OnGetCommand](https://github.com/Mimetis/Dotmim.Sync/discussions/802#discussioncomment-3594694)
- [TransactionMode](https://github.com/Mimetis/Dotmim.Sync/discussions/802#discussioncomment-3594703)

**Comment by Mimetis** [comment]
## Removing OnApplyChangesFailed with OnApplyChangesConflictOccured and OnApplyChangesErrorOccured

This version introduces the management of errors when changes are applied.
To be able to make the distinction between conflicts and errors, the `OnApplyChangesFailed` interceptor has been removed and replaced by

- `OnApplyChangesConflictOccured`: Occurs when a row generates conflict when applied
- `OnApplyChangesErrorOccured`: Occurs when a row generates an error when applied

### OnApplyChangesConflictOccured

Basically `OnApplyChangesConflictOccured` is replacing the old `OnApplyChangesFailed`.

This new feature add some new features and to improve the conflict handling, the conflict is not automatically retrieved from the underline data source.

You can get it using `GetSyncConflictAsync()`

``` cs
remoteOrchestrator.OnApplyChangesConflictOccured(async acf =>
{
    // Check conflict is correctly set
    var conflict = await acf.GetSyncConflictAsync();
    var localRow = conflict.LocalRow;
    var remoteRow = conflict.RemoteRow;
    
    // Client should wins
    acf.Resolution = ConflictResolution.ClientWins;
});
```
You have now a new enumeration `ConflictResolution.Throw`, that will mark the conflicting row as an error.
This row will be next intercepted by the `OnApplyChangesErrorOccured` interceptor.

### OnApplyChangesErrorOccured

Each time a row is in error (due to various reasons as **Unique constraint** failed or **ForeignKeys constraint** failed or any other failure that could happens), the row in error will be logged locally in a `BatchInfo` directory. Usually the batch info file has "**ERROR**" in its directory name.

`DMS` can handle the error row, depending on the situation and how you want to deal with it.
By default, any row in error will throw an `Exception`, and the transaction will be rollback.

But you can the way the error is handled:

A new interceptor has been implemented when an error has occurred during the `applyChanges` mechanism : `OnApplyChangesErrorOccured`.
You can choose different action depending on the row in error, and your scenario, using the `ErrorResolution` enumeration :

``` cs
/// <summary>
/// Determines what kind of action should be taken when an error is raised from the datasource
/// during an insert / update or delete command
/// </summary>
public enum ErrorResolution
{
    /// <summary>
    /// Ignore the error and continue to sync. Error will be stored 
    /// locally in a separate batch info file
    /// <para>
    /// Row is stored locally with a state of <see cref="SyncRowState.ApplyDeletedFailed"/> 
    /// or <see cref="SyncRowState.ApplyModifiedFailed"/> depending on the row state.
    /// </para>
    /// </summary>
    ContinueOnError,

    /// <summary>
    /// Will try one more time once after all the others rows in the table. 
    /// <para>
    /// If the error is raised again, it will be stored locally in a 
    /// separate batch info file with a state of <see cref="SyncRowState.ApplyDeletedFailed"/> 
    /// or <see cref="SyncRowState.ApplyModifiedFailed"/> depending on the row state.
    /// </para>
    /// </summary>
    RetryOneMoreTime,

    /// <summary>
    /// Row is stored locally and will be applied again on next sync.
    /// Row is stored locally with a state of <see cref="SyncRowState.RetryDeletedOnNextSync"/>
    /// or <see cref="SyncRowState.RetryModifiedOnNextSync"/> depending on the row state.
    /// </summary>
    RetryOnNextSync,

    /// <summary>
    /// Considers the row as applied.
    /// </summary>
    Resolved,

    /// <summary>
    /// Throw the error. Default value.
    /// </summary>
    Throw
}

```
Here is some examples with a foreign key constraint failure due to Parent column not available during the insert phase:

Simulating a Foreign Key failure, using this TSQL script:

``` sql
BEGIN TRAN
  ALTER TABLE [ProductCategory] NOCHECK CONSTRAINT ALL
    INSERT [ProductCategory] ([ProductCategoryID], [ParentProductCategoryId], [Name]) VALUES (N'A', 'B', N'A Sub category')
    INSERT [ProductCategory] ([ProductCategoryID], [ParentProductCategoryId], [Name]) VALUES (N'B', NULL, N'B Category');
  ALTER TABLE [ProductCategory] CHECK CONSTRAINT ALL
COMMIT TRAN;
```
> As you can see the **A** `ProductCategory` belongs to **B** `ProductCategory`. But because **A** is before **B**, on a Select statement, the **A** row will come firstly, before **B**. We have generated a **ForeignKey** constraint failure

By default, the sync is rollback, if we are making a sync, row by row 

> **Note**: Notice we are disabling **BulkOperation**, because using the bulk mode in Sql Server, a batch block is inserted in a single atomic operation, and the foreign key constraint exception will not be raised.

#### ErrorResolution.Throw

``` cs

// ErrorResolution.Throw is the default resolution. No need to explicitly set it.
// It's done here for the demo explanation.
agent.LocalOrchestrator.OnApplyChangesErrorOccured(args =>
{
    Console.ForegroundColor = ConsoleColor.Red;
    Console.WriteLine($"ERROR: {args.ErrorRow}");
    Console.ResetColor();
    args.Resolution = ErrorResolution.Throw;
});

var serverProvider = new SqlSyncProvider(DBHelper.GetDatabaseConnectionString("svProduct"));
serverProvider.UseBulkOperations = false;
var clientProvider = new SqlSyncProvider(DBHelper.GetDatabaseConnectionString("CliProduct"));
clientProvider.UseBulkOperations = false;
var setup = new SyncSetup("ProductCategory");
var options = new SyncOptions();

// Creating an agent that will handle all the process
var agent = new SyncAgent(clientProvider, serverProvider, options);

var result = await agent.SynchronizeAsync(setup);
Console.WriteLine(result);

```
![image](https://user-images.githubusercontent.com/4592555/189122927-ab0f0505-63dd-4487-a0c0-604a980544d3.png)

As you can see, error is raised, and sync is rollback. This is the classic behavior. 

You can see in the `Options.BatchDirectory` a new folder containing the errors logged:

![image](https://user-images.githubusercontent.com/4592555/189123365-f17677b1-e908-4468-b628-b60aa7f3ba19.png)

**This file is never deleted. Please check regularly your `DMS` tmp folder and purge logged errors if needed.**

By the way, you can load any (or all) batch info files using `LoadTableFromBatchInfoAsync()`, `LoadTablesFromBatchInfoAsync()` or `LoadBatchInfosAsync`

``` cs

// Creating an agent that will handle all the process
var agent = new SyncAgent(clientProvider, serverProvider, options);

try
{
  var result = await agent.SynchronizeAsync(setup);
  Console.WriteLine(result);
}
catch (Exception)
{
  // Loading all batch infos from tmp dir
  var batchInfos = await agent.LocalOrchestrator.LoadBatchInfosAsync();
  foreach (var batchInfo in batchInfos)
  {
      // Load all rows from error tables specifying the specific SyncRowState states
      var allTables = agent.LocalOrchestrator.LoadTablesFromBatchInfoAsync(batchInfo,
          SyncRowState.ApplyDeletedFailed | SyncRowState.ApplyModifiedFailed);

      // Enumerate all rows in error
      await foreach (var table in allTables)
          foreach (var row in table.Rows)
              Console.WriteLine(row);
  }
}
```
![image](https://user-images.githubusercontent.com/4592555/189126652-88ff99ff-2c21-49f0-89d6-ed43dabe688f.png)

#### ErrorResolution.RetryOneMoreTime

Second scenario: We are trying to apply **again one more time** the error row, using `ErrorResolution.RetryOneMoreTime`

``` cs
agent.LocalOrchestrator.OnApplyChangesErrorOccured(args =>
{
    Console.ForegroundColor = ConsoleColor.Red;
    Console.WriteLine($"ERROR: {args.ErrorRow}");
    Console.ResetColor();
    args.Resolution = ErrorResolution.RetryOneMoreTime;
});
```
![image](https://user-images.githubusercontent.com/4592555/189129176-8cd60b06-7bfe-4bc6-ac0a-2c32223a66c9.png)

As you can see in the previous screenshot, a Foreign key constraint prevents the rows to be inserted until the Parent category is inserted
Since we specified to **apply the rows once again after all the others rows have been applied**, the parent category is now available and then the rows are successfully applied at the end.

#### ErrorResolution.RetryOnNextSync

The idea here is to retry to sync the failed row (once again, stored locally) on the next sync.

``` cs
agent.LocalOrchestrator.OnApplyChangesErrorOccured(args =>
{
    Console.ForegroundColor = ConsoleColor.Red;
    Console.WriteLine($"ERROR: {args.ErrorRow}");
    Console.ResetColor();
    args.Resolution = ErrorResolution.RetryOnNextSync;
});

```
For the purpose of the demo, we are doing 2 sync in a row:

``` cs
// Creating an agent that will handle all the process
var agent = new SyncAgent(clientProvider, serverProvider, options);

Console.ResetColor();
Console.WriteLine("FIRST SYNC:");
var result = await agent.SynchronizeAsync(setup);
Console.WriteLine(result);

Console.WriteLine("SECOND SYNC:");
result = await agent.SynchronizeAsync(setup);
Console.WriteLine(result);

```
![image](https://user-images.githubusercontent.com/4592555/189130325-a7733c25-c98b-4faa-8f35-279b817f8e2a.png)

As you can see we have a first sync where we have a failed inserted row, that is applied on the next sync, even if we did not downloaded any rows from the server

#### ErrorResolution.Resolved

This resolution prevents the sync from failing and is assuming you are doing the necessaries actions to insert / update the row that has failed.
**TL;DR;**: It's ok for DMS, but do something !

``` cs
agent.LocalOrchestrator.OnApplyChangesErrorOccured(args =>
{
    Console.ForegroundColor = ConsoleColor.Red;
    Console.WriteLine($"ERROR: {args.ErrorRow}");
    Console.ResetColor();
    args.Resolution = ErrorResolution.Resolved;
});
```
![image](https://user-images.githubusercontent.com/4592555/189131023-e2790fca-051e-44e7-83c6-2502a5017b25.png)

As you explicitly said that the error row problem has been resolved, `DMS` is setting the row as "applied".

#### ErrorResolution.ContinueOnError

Almost same as `ErrorResolution.Resolved`, but DMS is considering the row as a failed applied row, but will continue to sync the next tables.

``` cs
agent.LocalOrchestrator.OnApplyChangesErrorOccured(args =>
{
    Console.ForegroundColor = ConsoleColor.Red;
    Console.WriteLine($"ERROR: {args.ErrorRow}");
    Console.ResetColor();
    args.Resolution = ErrorResolution.ContinueOnError;
});
```

![image](https://user-images.githubusercontent.com/4592555/189131643-4f9ea358-5ecb-4fa2-a82e-aacc6e51cc4c.png)

**Comment by Mimetis** [comment]
## Upgrade of scope_info tables

This version is a major upgrade version.
The scopes tables are changing, to be more flexible, and allow more scenario.

If you want to read a little bit on why `DMS `is changing its core tables structure, you can read the discussion #789 

Once upgraded, you will have 2 scopes tables on both sides

**Server**:
- `scope_info_server` table becomes `scope_info` with fewer fields
- `scope_info_history` tables becomes `scope_info_client`

**Client**
- `scope_info` table stays as `scope_info` with fewer fields
- New table  `scope_info_client` is created.

Basically, the two tables are identical on both sides.

- `scope_info`: This table now contains **only** the **setup** and **schema**. Each line is identified by the `scopeName` property.
- `scope_info_client`: This table contains all clients that needs to synchronize a setup from `scope_info`. 
Each client is identified by the `scopeName` property and the `syncParameters` used.

Those two tables are created on both side and are identical. 
- No more `scope_info_server` on server side (both sides have now a `scope_info` table)
- No more `scope_info_history` on server side (both sides have now a `scope_info_client` table)

As an example, here is a sync with 3 tables, 2 of them are filtered (`Product` and `ProductCategory`), the 3rd table does not have any filter (`Employee`)

Here is a `scope_info` containing a setup with 3 tables:
- `ProductCategory`
- `Product`
- `Employee`

<img width="752" alt="image" src="https://user-images.githubusercontent.com/4592555/187087016-fb28422c-a829-4f42-a35b-69de553977d2.png">

As you can see, we don't have any more the last sync timestamp information here.
The new `scope_info_client` contains 2 lines:
- First one is for the filter `ProductCategoryId == 'MOUNTB'`
- Second one is for the filter `ProductCategoryId == 'ROADFR`

Each `scope_info_client` record is totally independent and contains its own Last Sync Timestamp (server and client)

<img width="926" alt="image" src="https://user-images.githubusercontent.com/4592555/187087045-deda1eef-ac95-430d-87e9-0e5cb7ed3175.png">


Here is the code used to have this result:

``` cs
var setup = new SyncSetup("ProductCategory", "Product", "Employee");
setup.Tables[productCategoryTableName].Columns.AddRange("ProductCategoryId", "Name", "rowguid", "ModifiedDate");
setup.Filters.Add("ProductCategory", "ProductCategoryId");
setup.Filters.Add("Product", "ProductCategoryId");


var pMount = new SyncParameters(("ProductCategoryId", "MOUNTB"));
var pRoad = new SyncParameters(("ProductCategoryId", "ROADFR"));

var agent = new SyncAgent(client.Provider, Server.Provider);
var r1 = await agent.SynchronizeAsync("v1", setup, pMount);
var r2 = await agent.SynchronizeAsync("v1", setup, pRoad);
```

It's a big step moving forward for `DMS` as I never imagined to be able to sync a client database with different parameters value
Let me know what you think.

If you want to test this version, you can grab the branch [ClientHistory](https://github.com/Mimetis/Dotmim.Sync/tree/ClientHistory)

**Comment by Mimetis** [comment]
## Upgrade to v0.9.6

The upgrade is really tricky, especially from the client side.
Since this upgrade is quite complex, and I'm not sure to have anticipated all the scenario, you can specify a boolean value on the method `UpgradeAsync`, that will evaluate the migration, without applying it:

``` cs
remoteOrchestrator.UpgradeAsync(progress, evaluationOnly)
```

Do not hesitate to test the upgrade on both side (client (`LocalOrchestrator`) / server (`RemoteOrchestrator`))

### Server side

Pretty simple, you can make a classic upgrade as usual. Here is a detailed code to do so:

``` cs
var progress = new SynchronousProgress<ProgressArgs>(s =>
{
    Console.WriteLine($"{s.ProgressPercentage:p}:  
          \t[{s?.Source[..Math.Min(4, s.Source.Length)]}] {s.TypeName}: {s.Message}");
});

var serverProvider = new SqlSyncProvider(serverCstring);
var syncOptions = new SyncOptions();
var remoteOrchestrator = new RemoteOrchestrator(serverProvider, syncOptions);

var needsUpgrade = await remoteOrchestrator.NeedsToUpgradeAsync();

if (needsUpgrade)
{
    Console.WriteLine("----------------------");
    Console.WriteLine($"Evaluation only ? : {evaluationOnly}");
    Console.WriteLine("----------------------");

    var (scopeInfos, scopeInfoClients) = 
      await remoteOrchestrator.UpgradeAsync(progress, evaluationOnly);

    Console.WriteLine("----------------------");
    Console.WriteLine($"Upgrade done.");
    Console.WriteLine("----------------------");
    Console.WriteLine("Get Scope Infos from the new scope_info table:");
    foreach (var si in scopeInfos)
        Console.WriteLine(si);
    Console.WriteLine("----------------------");

    Console.WriteLine("----------------------");
    Console.WriteLine("Get Scope Info Clients from the new scope_info_client table (SHOULD BE EMPTY):");
    foreach (var sic in scopeInfoClients)
        Console.WriteLine(sic);
    Console.WriteLine("----------------------");

```
Here is the output of this code:

![image](https://user-images.githubusercontent.com/4592555/187503136-a5dce85a-d8d9-4849-98dc-fd1282fe7062.png)

### Client side

**First of all, the `UpgradeAsync()` is not called automatically. You need to call it manually**

Why ? Because I need you to call it with all your `SyncParameters` values.

The new **scope_info_client** table basically needs your `SyncParameters` values, that you are using on the client side when doing a synchronization.
These information were not saved before today.

That's why, for the client `UpgradeAsync`, I need you to fill all the `SyncParameters` you are using (and most important, the values of them)

I've created a special class called `ScopeInfoClientUpgrade` that will stored the scope name and the sync parameters values you are using.

As an example, imagine you are 
- Using a `SyncParameter ` instance containing a `ProductCategoryId` with a value of "`Guid("XXXX-XXXX")`"
- You are launching a sync with a scope name of "**v1**"

Here is the code to call to Upgrade to `v0.9.6`:

``` cs
var progress = new SynchronousProgress<ProgressArgs>(s =>
{
    Console.WriteLine($"{s.ProgressPercentage:p}:  " +
        $"\t[{s?.Source[..Math.Min(4, s.Source.Length)]}] {s.TypeName}: {s.Message}");
});

var clientProvider = new SqlSyncProvider(clientCstring);
var syncOptions = new SyncOptions();
var localOrchestrator = new LocalOrchestrator(clientProvider, syncOptions);

var needsUpgrade = await localOrchestrator.NeedsToUpgradeAsync();

if (needsUpgrade)
{
    var entries = new List<ScopeInfoClientUpgrade>();
    var entry = new ScopeInfoClientUpgrade
    {
        Parameters = new SyncParameters(("ProductCategoryId", new Guid("XXXX-XX"))),
        ScopeName = "v1"
    };
    entries.Add(entry);

    Console.WriteLine("----------------------");
    Console.WriteLine($"Evaluation only ? : {evaluationOnly}");
    Console.WriteLine("----------------------");

    var (scopeInfos, scopeInfoClients) = 
        await localOrchestrator.UpgradeAsync(entries, progress, evaluationOnly);

    Console.WriteLine("----------------------");
    Console.WriteLine($"Upgrade done.");
    Console.WriteLine("----------------------");
    Console.WriteLine("Get Scope Infos from the new scope_info table:");
    foreach (var si in scopeInfos)
        Console.WriteLine(si);
    Console.WriteLine("----------------------");

    Console.WriteLine("----------------------");
    Console.WriteLine("Get Scope Info Clients from the new scope_info_client table:");
    foreach (var sic in scopeInfoClients)
        Console.WriteLine(sic);
    Console.WriteLine("----------------------");

}

```
Here is the output of this code:

![image](https://user-images.githubusercontent.com/4592555/187502984-d101371c-5bc7-4226-aae5-edd25cb68906.png)

**Comment by Mimetis** [comment]
## OnGetCommand

`OnGetCommand` is raised each time a `DbCommand` is retrieved from the underline data source.
You can change whatever you want from the `DbCommand` if needed:

``` cs
// Eeach time a command is called on remote side, 
// We are making a redirection to "default" stored procedures, except for filters specific stored proc
agent.RemoteOrchestrator.OnGetCommand(args =>
{
    if (args.Command.CommandType == CommandType.StoredProcedure)
    {
        switch (args.CommandType)
        {
            case DbCommandType.SelectInitializedChangesWithFilters:
            case DbCommandType.SelectChangesWithFilters:
                break;
            default:
                args.Command.CommandText = args.Command.CommandText.Replace("_filterproducts_", "_default_");
                break;
        }
        Console.WriteLine(args.Command.CommandText);
    }
});
```
Take a look at the discussion #789 for a complete example

**Comment by Mimetis** [comment]
## TransactionMode

You can specify now a transaction mode when applying changes:

``` cs
    public enum TransactionMode
    {
        /// <summary>
        /// Default mode for transaction, when applying changes
        /// </summary>
        AllOrNothing,

        /// <summary>
        /// Each batch file will have its own transaction
        /// </summary>
        PerBatch,

        /// <summary>
        /// No transaction during applying changes. very risky
        /// </summary>
        None
    }
```
You can set the `TransactionMode` from the `SyncOptions` instance. By default `AllOrNothing` is used.
Be careful with this option and see discussion #793 for more context and why this option has been implemented.
